[jit] Add pass registration mechanism

bwasti opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #18846 [jit] Only require python print on certain namespaces
* #18588 [jit] Add Custom graph fusion
* #18589 [jit] Add options to Operator to enable registration of alias analysis passes
* **#18587 [jit] Add pass registration mechanism**

Differential Revision: [D14901227](https://our.internmc.facebook.com/intern/diff/D14901227)

bwasti committed:
[jit] Add pass registration mechanism

zdevito commented on torch/csrc/jit/pass_manager.h:
```diff
@@ -0,0 +1,30 @@
+namespace jit {
+
+// A pass modifies a Graph in place.
+typedef void (*Pass)(std::shared_ptr<Graph>&);
```
Probably better to use `std::function<void(std::shared_ptr<Graph>)>`. Otherwise registering things would have difficulty associating state with a pass.

zdevito commented on test/cpp/jit/test_misc.h:
```diff
@@ -643,6 +644,23 @@ void testNoneSchemaMatch() {
+
+RegisterPass p(fakePass);
+
+void testPassManagement() {
```
We should also have a test that the passes actually run when using the executor.

zdevito reviewed (approved):
Looks good, but should have a test that it actually runs.

ZolotukhinM commented on torch/csrc/jit/pass_manager.h:
```diff
@@ -0,0 +1,30 @@
+// A pass modifies a Graph in place.
+typedef void (*Pass)(std::shared_ptr<Graph>&);
+
+std::vector<Pass>& getPasses();
```
I must be missing something, but where is `Pass` defined?

bwasti replied:
2 lines above this comment (C style). This will be changed to C++ style using + std::function as zach suggested

ZolotukhinM commented on torch/csrc/jit/pass_manager.h:
```diff
@@ -0,0 +1,30 @@
+#include <torch/csrc/jit/ir.h>
+
+/* `getPasses()` returns a vector of passes that will be executed after
+ * differentiation but before any fusion.  This is the de-facto location
```
I wonder if it would make sense to generalize it a little and introduce 1) target-independent and 2) target-specific passes. The fusion would be in (2), but I expect more stuff than just fusion to happen there down the road.

bwasti replied:
Agreed, but I think we should push it to a later PR.  I believe there will be some other considerations when it comes to the hierarchy/order of pass invocations.

bwasti committed:
Update on "[jit] Add pass registration mechanism"

[jit] Add pass registration mechanism

gh-metadata: pytorch pytorch 18587 gh/bwasti/1/head

bwasti committed:
Update on "[jit] Add pass registration mechanism"

[jit] Add pass registration mechanism

gh-metadata: pytorch pytorch 18587 gh/bwasti/1/head

bwasti committed:
Update on "[jit] Add pass registration mechanism"

[jit] Add pass registration mechanism

gh-metadata: pytorch pytorch 18587 gh/bwasti/1/head

apaszke commented on torch/csrc/jit/pass_manager.cpp:
```diff
@@ -0,0 +1,16 @@
+namespace torch {
+namespace jit {
+
+std::vector<Pass>& getPasses() {
```
nit: can we make this slightly less general than `getPasses()`? More like `getCustomPasses()`?

bwasti committed:
Update on "[jit] Add pass registration mechanism"

[jit] Add pass registration mechanism

gh-metadata: pytorch pytorch 18587 gh/bwasti/1/head

wanchaol commented on torch/csrc/jit/pass_manager.cpp:
```diff
@@ -0,0 +1,16 @@
+RegisterPass::RegisterPass(Pass p) {
+  getPasses().emplace_back(std::move(p));
+}
+
```
shall we also expose an interface to remove a custom pass? it would be nice so that backend can control registering some passes and disable some custom passes they registered if necessary.

ZolotukhinM replied:
In traditional compilers is usually handled by having target hooks, which any pass can check. If some property is disabled on the current target, a pass just becomes no-op. That is, it remains in the pipeline, but its behavior depends on these target-specific properties. We can do something similar here I guess.

bwasti committed:
Update on "[jit] Add pass registration mechanism"

[jit] Add pass registration mechanism

gh-metadata: pytorch pytorch 18587 gh/bwasti/1/head

bwasti committed:
Update on "[jit] Add pass registration mechanism"

[jit] Add pass registration mechanism

gh-metadata: pytorch pytorch 18587 gh/bwasti/1/head

bwasti committed:
Update on "[jit] Add pass registration mechanism"

[jit] Add pass registration mechanism

gh-metadata: pytorch pytorch 18587 gh/bwasti/1/head
